File: bunood_theme/public/js/finance_close.js
// Copyright (c) 2026, Bunood and contributors
// A read-first comprehension layer over ERPNext's native close controls.

/* eslint-env browser */
/* global frappe, __ */

(function () {
	"use strict";

	const METHOD = "bunood_theme.api.finance_close_cockpit";

	function el(tag, className, text) {
		const node = document.createElement(tag);
		if (className) node.className = className;
		if (text !== undefined && text !== null) node.textContent = text;
		return node;
	}

	function button(label, route, primary) {
		const node = el("button", `btn ${primary ? "btn-primary" : "btn-default"}`, label);
		node.type = "button";
		node.addEventListener("click", () => frappe.set_route(...route));
		return node;
	}

	function control(host, options) {
		const field = frappe.ui.form.make_control({ parent: host, df: options, render_input: true });
		field.refresh();
		return field;
	}

	function stateCopy(state) {
		return {
			"attention-needed": [__("Attention needed"), __("Draft or failed native records need review before this period can move forward.")],
			"incomplete-evidence": [__("Evidence incomplete"), __("One or more permitted queries could not be completed. No close conclusion was inferred.")],
			"protection-present": [__("Protection is present"), __("A native posting restriction covers this period. Reconciliations and approval still require review.")],
			"review-required": [__("Review required"), __("No period protection was observed for this boundary. Complete review before applying a lock.")],
		}[state] || [__("Review required"), __("Review the native records for this period.")];
	}

	function metric(label, value, detail, warning) {
		const card = el("div", `bnd-close__metric${warning ? " is-warning" : ""}`);
		card.append(el("span", "", label), el("strong", "", String(value)), el("small", "", detail));
		return card;
	}

	function controlCard(title, value, note, active) {
		const card = el("article", `bnd-close__control${active ? " is-present" : ""}`);
		const head = el("div", "bnd-close__control-head");
		head.append(el("h3", "", title), el("span", "bnd-close__status", active ? __("Observed") : __("Not observed")));
		card.append(head, el("strong", "", value || __("None for this period")), el("p", "", note));
		return card;
	}

	function renderAttention(data) {
		const panel = el("section", "bnd-close__panel");
		panel.append(el("h2", "", __("Attention now")));
		if (!data.attention.length) {
			panel.append(el("p", "bnd-close__empty", __("No draft or failed native records were observed in the selected boundary. This is not close approval.")));
			return panel;
		}
		const wrap = el("div", "bnd-close__table-wrap");
		const table = el("table", "bnd-close__table");
		const head = el("tr");
		[__("Document"), __("Type"), __("Date"), __("Reason")].forEach((label) => head.append(el("th", "", label)));
		const thead = el("thead");
		thead.append(head);
		const body = el("tbody");
		data.attention.forEach((row) => {
			const tr = el("tr");
			const linkCell = el("td");
			const link = el("button", "bnd-close__link", row.name);
			link.type = "button";
			link.addEventListener("click", () => frappe.set_route("Form", row.doctype, row.name));
			linkCell.append(link);
			tr.append(linkCell, el("td", "", __(row.doctype)), el("td", "", row.date || "—"), el("td", "", row.reason === "failed-closing-processing" ? __("Closing processing failed") : __("Draft source document")));
			body.append(tr);
		});
		table.append(thead, body);
		wrap.append(table);
		panel.append(wrap);
		if (data.attention_truncated) panel.append(el("p", "bnd-close__note", __("The queue is limited. Open the native lists to review all records.")));
		return panel;
	}

	function renderPhases(data) {
		const names = {
			"source-capture": [__("1. Source capture"), __("Draft transactions and cutoff")],
			reconciliation: [__("2. Reconciliations"), __("Bank, AR/AP, stock, assets and tax")],
			adjustments: [__("3. Adjustments"), __("Journals, schedules and review")],
			statements: [__("4. Statements"), __("Trial Balance and financial statements")],
			protection: [__("5. Period protection"), __("Native restriction after approval")],
		};
		const panel = el("section", "bnd-close__panel");
		panel.append(el("h2", "", __("Close path")), el("p", "bnd-close__note", __("These are evidence stages, not completion badges. Not evaluated remains different from complete.")));
		const grid = el("div", "bnd-close__phases");
		data.phases.forEach((phase) => {
			const card = el("article", `bnd-close__phase is-${phase.state}`);
			card.append(el("h3", "", names[phase.id][0]), el("p", "", names[phase.id][1]), el("span", "", phase.state === "attention-needed" ? __("Attention needed") : phase.state === "protection-present" ? __("Protection observed") : phase.state === "not-protected" ? __("Not protected") : __("Not evaluated")));
			grid.append(card);
		});
		panel.append(grid);
		return panel;
	}

	function renderReports(data) {
		const panel = el("section", "bnd-close__panel");
		panel.append(el("h2", "", __("Review reports")), el("p", "bnd-close__note", __("Each report opens with native permissions. Apply the same company, dates, currency, Finance Book and dimensions before comparing figures.")));
		const actions = el("div", "bnd-close__report-grid");
		[
			[__("Trial Balance"), ["query-report", "Trial Balance"]],
			[__("General Ledger"), ["query-report", "General Ledger"]],
			[__("Profit and Loss Statement"), ["query-report", "Profit and Loss Statement"]],
			[__("Balance Sheet"), ["query-report", "Balance Sheet"]],
			[__("Cash Flow"), ["query-report", "Cash Flow"]],
			[__("Accounts Receivable"), ["query-report", "Accounts Receivable"]],
			[__("Accounts Payable"), ["query-report", "Accounts Payable"]],
		].forEach(([label, route]) => actions.append(button(label, route, false)));
		if (!data.capabilities.can_read_gl) actions.setAttribute("aria-disabled", "true");
		panel.append(actions);
		return panel;
	}

	function renderResult(data, content) {
		content.replaceChildren();
		const state = el("section", `bnd-close__state is-${data.state}`);
		const copy = stateCopy(data.state);
		state.setAttribute("role", "status");
		state.setAttribute("aria-live", "polite");
		state.append(el("div", "", copy[0]), el("p", "", copy[1]), el("small", "", `${data.company} · ${data.from_date} — ${data.to_date} · ${__("Generated")} ${data.generated_at}`));

		const metrics = el("section", "bnd-close__metrics");
		metrics.append(
			metric(__("Items requiring attention"), data.summary.attention_count, __("Drafts and failed native records"), data.summary.attention_count > 0),
			metric(__("Draft source documents"), data.summary.draft_source_count, __("Inside the selected period"), data.summary.draft_source_count > 0),
			metric(__("Covering accounting periods"), data.summary.accounting_period_count, __("Native selective restrictions"), false),
			metric(__("Submitted closing vouchers"), data.summary.submitted_closing_voucher_count, __("P&L transfer; not a lock"), false)
		);

		const controls = el("section", "bnd-close__controls");
		controls.append(
			controlCard(__("Accounts frozen through"), data.controls.accounts_frozen_till_date || "—", __("Broad Company posting cut-off; authorised roles may bypass it."), data.controls.frozen_through_period),
			controlCard(__("Accounting Period"), data.controls.accounting_period_covers_period ? __("Covers selected boundary") : "—", __("Restricts configured document types; it does not transfer profit or loss."), data.controls.accounting_period_covers_period),
			controlCard(__("Period Closing Voucher"), data.controls.closing_voucher_covers_period ? __("Submitted voucher observed") : "—", __("Transfers Profit and Loss to the selected closing account; it does not lock posting."), data.controls.closing_voucher_covers_period)
		);

		content.append(state, metrics, controls, renderAttention(data), renderPhases(data), renderReports(data));
	}

	function render(container) {
		container.replaceChildren();
		const root = el("main", "bnd-close");
		const intro = el("header", "bnd-close__intro");
		const introCopy = el("div", "bnd-close__intro-copy");
		introCopy.append(el("span", "bnd-close__eyebrow", __("Native ERPNext controls")), el("h1", "", __("Finance and close")), el("p", "", __("See what needs attention, review the same reporting boundary, then use ERPNext's native controls. Bunood does not certify the books.")));
		const actions = el("div", "bnd-close__actions");
		actions.append(button(__("Journal workbench"), ["bnd-journal-workbench"], false), button(__("Accounting Periods"), ["List", "Accounting Period"], false), button(__("Closing Vouchers"), ["List", "Period Closing Voucher"], false));
		intro.append(introCopy, actions);

		const filters = el("section", "bnd-close__filters");
		const companyHost = el("div", "bnd-close__field");
		const fromHost = el("div", "bnd-close__field");
		const toHost = el("div", "bnd-close__field");
		const applyHost = el("div", "bnd-close__apply");
		const apply = el("button", "btn btn-primary", __("Review period"));
		apply.type = "button";
		applyHost.append(apply);
		filters.append(companyHost, fromHost, toHost, applyHost);
		const content = el("div", "bnd-close__content");
		root.append(intro, filters, content);
		container.append(root);

		const company = control(companyHost, { fieldtype: "Link", fieldname: "company", label: __("Company"), options: "Company", reqd: 1 });
		const from = control(fromHost, { fieldtype: "Date", fieldname: "from_date", label: __("From date"), reqd: 1 });
		const to = control(toHost, { fieldtype: "Date", fieldname: "to_date", label: __("To date"), reqd: 1 });
		company.set_value(frappe.defaults.get_user_default("Company") || "");
		to.set_value(frappe.datetime.get_today());
		from.set_value(frappe.datetime.month_start());

		const load = () => {
			if (!company.get_value()) {
				frappe.msgprint(__("Company is required"));
				return;
			}
			root.classList.add("is-loading");
			apply.disabled = true;
			frappe.call({
				method: METHOD,
				args: { company: company.get_value(), from_date: from.get_value(), to_date: to.get_value() },
			}).then((response) => renderResult(response.message, content)).catch(() => {
				const failure = el("section", "bnd-close__failure");
				failure.setAttribute("role", "alert");
				failure.append(el("h2", "", __("The close evidence could not be loaded")), el("p", "", __("Your controls are unchanged. Review the error, then try again.")));
				content.replaceChildren(failure);
			}).finally(() => {
				root.classList.remove("is-loading");
				apply.disabled = false;
			});
		};
		apply.addEventListener("click", load);
		if (company.get_value()) load();
	}

	window.bunood_theme = window.bunood_theme || {};
	window.bunood_theme.finance_close_render = render;
})();

File: bunood_theme/finance_close.py
# Copyright (c) 2026, Bunood and contributors
# For license information, please see license.txt
"""Permission-safe evidence for Bunood's finance-close cockpit.

The cockpit coordinates native ERPNext records. It does not maintain balances,
approve reconciliations, lock periods, or infer that the books are correct.
"""

from __future__ import annotations

from frappe.utils import date_diff, getdate, now_datetime, nowdate

import frappe


MAX_PERIOD_DAYS = 366
QUEUE_LIMIT = 60


def _can(doctype: str, permission: str) -> bool:
    return bool(
        frappe.db.exists("DocType", doctype)
        and frappe.has_permission(doctype, permission)
    )


def _period(from_date=None, to_date=None):
    end = getdate(to_date or nowdate())
    start = getdate(from_date or end.replace(day=1))
    days = date_diff(end, start)
    if days < 0:
        frappe.throw(frappe._("From date must be before or equal to To date"))
    if days > MAX_PERIOD_DAYS:
        frappe.throw(
            frappe._("Maximum close review period in days: {0}").format(
                MAX_PERIOD_DAYS
            )
        )
    return start, end


def _safe_list(doctype: str, *, errors: list[str], **kwargs) -> list[dict]:
    """Query through Frappe permissions and preserve unknown/error as evidence."""

    if not _can(doctype, "read"):
        return []
    try:
        return frappe.get_list(doctype, **kwargs)
    except Exception:
        errors.append(doctype)
        frappe.log_error(title=f"bunood_theme: finance close {doctype} query stood down")
        return []


def _attention_queue(company: str, start, end, errors: list[str]) -> tuple[list[dict], dict]:
    specs = (
        ("Journal Entry", "posting_date"),
        ("Sales Invoice", "posting_date"),
        ("Purchase Invoice", "posting_date"),
        ("Payment Entry", "posting_date"),
    )
    rows = []
    counts = {}
    for doctype, date_field in specs:
        if not _can(doctype, "read"):
            counts[doctype] = None
            continue
        records = _safe_list(
            doctype,
            errors=errors,
            filters={
                "company": company,
                "docstatus": 0,
                date_field: ["between", [str(start), str(end)]],
            },
            fields=["name", date_field, "modified"],
            order_by=f"{date_field} desc, modified desc",
            limit=0,
        )
        counts[doctype] = len(records)
        for record in records[:QUEUE_LIMIT]:
            rows.append(
                {
                    "doctype": doctype,
                    "name": record.get("name"),
                    "date": str(record.get(date_field) or ""),
                    "reason": "draft-source-document",
                }
            )
    rows.sort(key=lambda row: (row.get("date") or "", row.get("name") or ""), reverse=True)
    return rows[:QUEUE_LIMIT], counts


def _covers(row: dict, start, end) -> bool:
    return bool(
        row.get("start_date")
        and row.get("end_date")
        and getdate(row["start_date"]) <= start
        and getdate(row["end_date"]) >= end
    )


def _pcv_covers(row: dict, start, end) -> bool:
    return bool(
        row.get("period_start_date")
        and row.get("period_end_date")
        and getdate(row["period_start_date"]) <= start
        and getdate(row["period_end_date"]) >= end
    )


def get_finance_close_cockpit(company: str, from_date=None, to_date=None) -> dict:
    """Return close evidence without upgrading observations into approval."""

    if not company:
        frappe.throw(frappe._("Company is required"))
    company_doc = frappe.get_doc("Company", company)
    company_doc.check_permission("read")
    start, end = _period(from_date, to_date)
    errors: list[str] = []

    attention, draft_counts = _attention_queue(company, start, end, errors)
    periods = _safe_list(
        "Accounting Period",
        errors=errors,
        filters=[
            ["company", "=", company],
            ["disabled", "=", 0],
            ["start_date", "<=", str(end)],
            ["end_date", ">=", str(start)],
        ],
        fields=["name", "start_date", "end_date", "exempted_role", "modified"],
        order_by="end_date desc, modified desc",
        limit=20,
    )
    vouchers = _safe_list(
        "Period Closing Voucher",
        errors=errors,
        filters=[
            ["company", "=", company],
            ["period_end_date", ">=", str(start)],
            ["period_start_date", "<=", str(end)],
        ],
        fields=[
            "name",
            "docstatus",
            "period_start_date",
            "period_end_date",
            "closing_account_head",
            "gle_processing_status",
            "modified",
        ],
        order_by="period_end_date desc, modified desc",
        limit=20,
    )

    failed_vouchers = [
        row for row in vouchers if row.get("gle_processing_status") == "Failed"
    ]
    for row in failed_vouchers:
        attention.append(
            {
                "doctype": "Period Closing Voucher",
                "name": row.get("name"),
                "date": str(row.get("period_end_date") or ""),
                "reason": "failed-closing-processing",
            }
        )

    frozen_date = company_doc.get("accounts_frozen_till_date")
    frozen_through = bool(frozen_date and getdate(frozen_date) >= end)
    covering_periods = [row for row in periods if _covers(row, start, end)]
    covering_vouchers = [
        row
        for row in vouchers
        if row.get("docstatus") == 1 and _pcv_covers(row, start, end)
    ]
    protection_present = bool(frozen_through or covering_periods)
    draft_total = sum(value for value in draft_counts.values() if value is not None)

    if errors:
        state = "incomplete-evidence"
    elif attention:
        state = "attention-needed"
    elif protection_present:
        state = "protection-present"
    else:
        state = "review-required"

    return {
        "company": company,
        "currency": company_doc.get("default_currency"),
        "finance_book": company_doc.get("default_finance_book"),
        "from_date": str(start),
        "to_date": str(end),
        "generated_at": str(now_datetime()),
        "state": state,
        "query_errors": errors,
        "capabilities": {
            "can_read_gl": _can("GL Entry", "read"),
            "can_create_journal": _can("Journal Entry", "create"),
            "can_create_accounting_period": _can("Accounting Period", "create"),
            "can_create_closing_voucher": _can("Period Closing Voucher", "create"),
            "can_change_company": _can("Company", "write"),
        },
        "summary": {
            "draft_source_count": draft_total,
            "attention_count": len(attention),
            "accounting_period_count": len(covering_periods),
            "submitted_closing_voucher_count": len(covering_vouchers),
        },
        "draft_counts": draft_counts,
        "controls": {
            "accounts_frozen_till_date": str(frozen_date or ""),
            "frozen_through_period": frozen_through,
            "accounting_period_covers_period": bool(covering_periods),
            "closing_voucher_covers_period": bool(covering_vouchers),
            "protection_present": protection_present,
        },
        "attention": attention[:QUEUE_LIMIT],
        "attention_truncated": len(attention) > QUEUE_LIMIT,
        "accounting_periods": periods,
        "closing_vouchers": vouchers,
        "phases": [
            {
                "id": "source-capture",
                "state": "attention-needed" if draft_total else "not-evaluated",
            },
            {"id": "reconciliation", "state": "not-evaluated"},
            {
                "id": "adjustments",
                "state": "attention-needed"
                if (draft_counts.get("Journal Entry") or 0)
                else "not-evaluated",
            },
            {"id": "statements", "state": "not-evaluated"},
            {
                "id": "protection",
                "state": "protection-present" if protection_present else "not-protected",
            },
        ],
    }
